File: cli/lib/review-handoff.mjs
import { lstatSync, readFileSync, realpathSync } from "node:fs";
import path from "node:path";

import {
  canonicalReviewHandoffStages,
  reviewHandoffArtifact,
  reviewHandoffPendingApprovals,
  validateReviewHandoff,
} from "./review-handoff-contract.mjs";
import { reviewWorkflowDigest } from "./review-workflow-contract.mjs";
import { unknownOptionMessage } from "./suggest.mjs";

const MAX_JSON_BYTES = 5 * 1024 * 1024;
const REVIEW_HANDOFF_OPTIONS = [
  "-h",
  "--help",
  "--recipient",
  "--quality-report",
  "--browser-verification",
  "--json",
];

function optionValue(args, index, option) {
  const value = args[index + 1];
  if (!value || value.startsWith("--")) throw new Error(`${option} requires a value`);
  return value;
}

function requiredText(value, field) {
  if (typeof value !== "string" || value.trim() === "") {
    throw new Error(`${field} must be a non-empty string`);
  }
  return value.trim();
}

function optionalSource(value, field) {
  if (value === undefined || value === null || value === "") return "";
  if (typeof value !== "string" || value.trim() === "") {
    throw new Error(`${field} must be a non-empty JSON string`);
  }
  return value;
}

function parseJson(source, field) {
  try {
    return JSON.parse(source);
  } catch {
    throw new Error(`${field} must be valid JSON`);
  }
}

export function parseReviewHandoffArgs(args) {
  const parsed = {
    workflowPath: "",
    recipient: "",
    qualityReportPath: "",
    browserVerificationPath: "",
    json: false,
    help: false,
  };

  const options = new Map([
    ["--recipient", "recipient"],
    ["--quality-report", "qualityReportPath"],
    ["--browser-verification", "browserVerificationPath"],
  ]);
  for (let index = 0; index < args.length; index += 1) {
    const arg = args[index];
    if (arg === "-h" || arg === "--help") parsed.help = true;
    else if (arg === "--json") parsed.json = true;
    else if (options.has(arg)) {
      parsed[options.get(arg)] = optionValue(args, index, arg);
      index += 1;
    } else if (arg.startsWith("-")) {
      throw new Error(unknownOptionMessage("review-handoff", arg, REVIEW_HANDOFF_OPTIONS));
    } else if (!parsed.workflowPath) parsed.workflowPath = arg;
    else throw new Error(`review-handoff accepts one review workflow; received unexpected argument: ${arg}`);
  }
  return parsed;
}

export function readReviewHandoffInput(filePath, cwd = process.cwd(), field = "review handoff input") {
  if (!filePath) throw new Error(`${field} requires a JSON file`);
  const resolved = path.resolve(cwd, filePath);
  const stat = lstatSync(resolved);
  if (!stat.isFile() || stat.isSymbolicLink()) {
    throw new Error(`${field} must be a regular non-symbolic-link file`);
  }
  if (stat.size > MAX_JSON_BYTES) throw new Error(`${field} exceeds the 5 MB limit`);
  const reference = realpathSync(resolved);
  return { source: readFileSync(reference, "utf8"), reference };
}

export function buildReviewHandoff(workflowSource, options = {}) {
  const workflowJson = optionalSource(workflowSource, "workflowSource");
  const recipient = requiredText(options.recipient, "recipient");
  const workflowRef = requiredText(options.workflowRef, "workflowRef");
  const qualityReportSource = optionalSource(options.qualityReportSource, "qualityReportSource");
  const browserVerificationSource = optionalSource(
    options.browserVerificationSource,
    "browserVerificationSource",
  );
  const hasBrowserVerification = Boolean(browserVerificationSource);
  if (Boolean(qualityReportSource) !== hasBrowserVerification) {
    throw new Error("qualityReportSource and browserVerificationSource must be supplied together");
  }

  const workflow = parseJson(workflowJson, "workflowSource");
  const qualityReport = hasBrowserVerification
    ? parseJson(qualityReportSource, "qualityReportSource")
    : null;
  const browserVerification = hasBrowserVerification
    ? parseJson(browserVerificationSource, "browserVerificationSource")
    : null;
  const artifacts = {
    reviewWorkflow: reviewHandoffArtifact(workflowJson, workflow, workflowRef),
    qualityReport: hasBrowserVerification
      ? reviewHandoffArtifact(
        qualityReportSource,
        qualityReport,
        requiredText(options.qualityReportRef, "qualityReportRef"),
      )
      : null,
    browserVerification: hasBrowserVerification
      ? reviewHandoffArtifact(
        browserVerificationSource,
        browserVerification,
        requiredText(options.browserVerificationRef, "browserVerificationRef"),
      )
      : null,
  };

  const handoff = {
    kind: "design-ai-review-handoff",
    schemaVersion: 1,
    status: hasBrowserVerification ? "browser-evidence-prepared" : "static-evidence-prepared",
    recipient: {
      name: recipient,
      delivery: "not-delivered",
      consumerValidation: "pending",
    },
    artifacts,
    linkage: {
      status: "pass",
      reviewWorkflowArtifactSha256: reviewWorkflowDigest(workflow),
      qualityReportArtifactSha256: workflow.linkage?.reportSha256 || "",
      browserVerificationArtifactSha256: browserVerification
        ? reviewWorkflowDigest(browserVerification)
        : null,
      qualityReportArtifactMatch: browserVerification ? true : null,
      browserSourceReportMatch: browserVerification ? true : null,
      viewportCoverage: browserVerification ? "pass" : "not-run",
    },
    stages: canonicalReviewHandoffStages(hasBrowserVerification),
    nextAction: {
      id: "consumer-validation-required",
      status: "pending",
      summary: `Deliver this prepared handoff to ${recipient}, then validate it before implementation.`,
      approvalRequiredBefore: reviewHandoffPendingApprovals(
        workflow,
        browserVerification?.summary.status === "pass",
      ),
    },
    boundary: {
      mode: "read-only",
      localWrites: false,
      targetRepoMutation: false,
      externalWrites: false,
      deliveryPerformed: false,
    },
  };
  return validateReviewHandoff(handoff);
}

export function buildReviewHandoffFromFiles(parsed, cwd = process.cwd()) {
  if (!parsed.workflowPath) throw new Error("review-handoff requires a review workflow JSON file");
  if (!parsed.recipient) throw new Error("review-handoff requires --recipient");
  const workflow = readReviewHandoffInput(parsed.workflowPath, cwd, "review workflow");
  const hasQualityReport = Boolean(parsed.qualityReportPath);
  const hasBrowserVerification = Boolean(parsed.browserVerificationPath);
  if (hasQualityReport !== hasBrowserVerification) {
    throw new Error("--quality-report and --browser-verification must be supplied together");
  }
  const qualityReport = hasQualityReport
    ? readReviewHandoffInput(parsed.qualityReportPath, cwd, "quality report")
    : null;
  const browserVerification = hasBrowserVerification
    ? readReviewHandoffInput(parsed.browserVerificationPath, cwd, "browser verification")
    : null;
  return buildReviewHandoff(workflow.source, {
    workflowRef: workflow.reference,
    recipient: parsed.recipient,
    qualityReportSource: qualityReport?.source,
    qualityReportRef: qualityReport?.reference,
    browserVerificationSource: browserVerification?.source,
    browserVerificationRef: browserVerification?.reference,
  });
}

export function renderReviewHandoffMarkdown(handoff) {
  const lines = [
    `# Review handoff for ${handoff.recipient.name}`,
    "",
    `- Status: ${handoff.status}`,
    `- Delivery: ${handoff.recipient.delivery}`,
    `- Consumer validation: ${handoff.recipient.consumerValidation}`,
    `- Linkage: ${handoff.linkage.status}`,
    `- Boundary: ${handoff.boundary.mode}; no local write, target-repository mutation, external write, or delivery`,
    "",
    "## Evidence",
    "",
    `- Review workflow: ${handoff.artifacts.reviewWorkflow.reference}`,
    handoff.artifacts.qualityReport
      ? `- Quality report: ${handoff.artifacts.qualityReport.reference}`
      : "- Quality report: nested in the review workflow; no browser source file supplied",
    handoff.artifacts.browserVerification
      ? `- Browser verification: ${handoff.artifacts.browserVerification.reference}`
      : "- Browser verification: not run",
    "",
    "## Stages",
    "",
    ...handoff.stages.map((stage) => `- ${stage.id}: ${stage.status}`),
    "",
    "## Next action",
    "",
    `- ${handoff.nextAction.summary}`,
    ...handoff.nextAction.approvalRequiredBefore.map(
      (requirement) => `- Approval required before ${requirement}.`,
    ),
  ];
  return lines.join("\n");
}

File: tools/audit/smoke_domains/review_handoff.py
"""P7 handoff and P8 receipt smoke assertions."""
from __future__ import annotations

import hashlib
import json
from pathlib import Path

from .assertion_helpers import assert_no_ansi, assert_smoke_json_keys
from .review_quality import review_workflow_digest


def assert_review_handoff_json(
    raw: str,
    workflow_path: Path,
    *,
    recipient: str,
    context: str,
    cmd: list[str],
) -> None:
    assert_no_ansi(raw, cmd)
    try:
        handoff = json.loads(raw)
    except json.JSONDecodeError as error:
        raise SystemExit(f"failed to parse review handoff JSON after {context}") from error

    assert_smoke_json_keys(
        handoff,
        [
            "kind", "schemaVersion", "status", "recipient", "artifacts",
            "linkage", "stages", "nextAction", "boundary",
        ],
        label="top-level",
        context=context,
        command_label="review handoff JSON",
    )
    if (
        handoff.get("kind") != "design-ai-review-handoff"
        or handoff.get("schemaVersion") != 1
        or handoff.get("status") != "static-evidence-prepared"
    ):
        raise SystemExit(f"review handoff JSON after {context} changed its contract identity")
    if handoff.get("recipient") != {
        "name": recipient,
        "delivery": "not-delivered",
        "consumerValidation": "pending",
    }:
        raise SystemExit(f"review handoff JSON after {context} changed its pending recipient state")

    artifacts = handoff.get("artifacts")
    if not isinstance(artifacts, dict) or set(artifacts) != {
        "reviewWorkflow", "qualityReport", "browserVerification",
    }:
        raise SystemExit(f"review handoff JSON after {context} changed its artifact inventory")
    workflow_artifact = artifacts.get("reviewWorkflow")
    workflow_source = workflow_path.read_text(encoding="utf-8")
    if not isinstance(workflow_artifact, dict) or set(workflow_artifact) != {
        "reference", "sha256", "bytes", "source", "value",
    }:
        raise SystemExit(f"review handoff JSON after {context} changed its workflow artifact shape")
    if (
        workflow_artifact.get("reference") != str(workflow_path.resolve())
        or workflow_artifact.get("source") != workflow_source
        or workflow_artifact.get("sha256") != hashlib.sha256(workflow_source.encode("utf-8")).hexdigest()
        or workflow_artifact.get("bytes") != len(workflow_source.encode("utf-8"))
    ):
        raise SystemExit(f"review handoff JSON after {context} lost exact workflow source identity")
    try:
        workflow_value = json.loads(workflow_source)
    except json.JSONDecodeError as error:
        raise SystemExit(f"review handoff source workflow after {context} is invalid JSON") from error
    if workflow_artifact.get("value") != workflow_value:
        raise SystemExit(f"review handoff JSON after {context} changed its embedded workflow value")
    if artifacts.get("qualityReport") is not None or artifacts.get("browserVerification") is not None:
        raise SystemExit(f"review handoff JSON after {context} added undeclared browser evidence")

    linkage = handoff.get("linkage")
    expected_linkage = {
        "status": "pass",
        "reviewWorkflowArtifactSha256": review_workflow_digest(workflow_value),
        "qualityReportArtifactSha256": workflow_value.get("linkage", {}).get("reportSha256"),
        "browserVerificationArtifactSha256": None,
        "qualityReportArtifactMatch": None,
        "browserSourceReportMatch": None,
        "viewportCoverage": "not-run",
    }
    if linkage != expected_linkage:
        raise SystemExit(f"review handoff JSON after {context} changed artifact linkage evidence")
    if handoff.get("stages") != [
        {"id": "plan", "status": "complete", "artifactKind": "design-ai-start"},
        {"id": "static-review", "status": "complete", "artifactKind": "design-ai-quality-report"},
        {"id": "browser-verification", "status": "not-run", "artifactKind": None},
        {"id": "implementation-handoff", "status": "prepared", "artifactKind": "design-ai-review-handoff"},
    ]:
        raise SystemExit(f"review handoff JSON after {context} changed its stage sequence")
    if handoff.get("nextAction") != {
        "id": "consumer-validation-required",
        "status": "pending",
        "summary": f"Deliver this prepared handoff to {recipient}, then validate it before implementation.",
        "approvalRequiredBefore": workflow_value.get("nextAction", {}).get("approvalRequiredBefore"),
    }:
        raise SystemExit(f"review handoff JSON after {context} changed its consumer gate")
    if handoff.get("boundary") != {
        "mode": "read-only",
        "localWrites": False,
        "targetRepoMutation": False,
        "externalWrites": False,
        "deliveryPerformed": False,
    }:
        raise SystemExit(f"review handoff JSON after {context} changed its undelivered boundary")


def assert_review_handoff_receipt_json(
    raw: str,
    handoff_path: Path,
    *,
    consumer: str,
    context: str,
    cmd: list[str],
) -> None:
    assert_no_ansi(raw, cmd)
    try:
        receipt = json.loads(raw)
    except json.JSONDecodeError as error:
        raise SystemExit(f"failed to parse review handoff receipt JSON after {context}") from error

    assert_smoke_json_keys(
        receipt,
        [
            "kind", "schemaVersion", "status", "consumer", "handoff", "evidence",
            "remainingApprovals", "nextAction", "boundary",
        ],
        label="top-level",
        context=context,
        command_label="review handoff receipt JSON",
    )
    if (
        receipt.get("kind") != "design-ai-review-handoff-receipt"
        or receipt.get("schemaVersion") != 1
        or receipt.get("status") != "contract-validated"
    ):
        raise SystemExit(f"review handoff receipt JSON after {context} changed its contract identity")

    handoff_source = handoff_path.read_text(encoding="utf-8")
    try:
        handoff_value = json.loads(handoff_source)
    except json.JSONDecodeError as error:
        raise SystemExit(f"review handoff receipt source after {context} is invalid JSON") from error
    if receipt.get("consumer") != {
        "name": consumer,
        "expectedRecipient": consumer,
        "recipientMatch": True,
        "identity": "self-declared",
        "contractValidation": "pass",
        "acceptance": "not-claimed",
    }:
        raise SystemExit(f"review handoff receipt JSON after {context} changed its consumer boundary")
    handoff_artifact = receipt.get("handoff")
    if not isinstance(handoff_artifact, dict) or set(handoff_artifact) != {
        "reference", "sha256", "bytes", "source", "value",
    }:
        raise SystemExit(f"review handoff receipt JSON after {context} changed its handoff artifact")
    if (
        handoff_artifact.get("reference") != str(handoff_path.resolve())
        or handoff_artifact.get("source") != handoff_source
        or handoff_artifact.get("value") != handoff_value
        or handoff_artifact.get("sha256") != hashlib.sha256(handoff_source.encode("utf-8")).hexdigest()
        or handoff_artifact.get("bytes") != len(handoff_source.encode("utf-8"))
    ):
        raise SystemExit(f"review handoff receipt JSON after {context} lost exact handoff identity")

    report_summary = handoff_value["artifacts"]["reviewWorkflow"]["value"]["report"]["summary"]
    browser_artifact = handoff_value["artifacts"]["browserVerification"]
    browser_status = browser_artifact["value"]["summary"]["status"] if browser_artifact else "not-run"
    if receipt.get("evidence") != {
        "qualityStatus": report_summary["status"],
        "confirmedFindings": report_summary["confirmedFindings"],
        "unverifiedFindings": report_summary["unverifiedFindings"],
        "browserStatus": browser_status,
    }:
        raise SystemExit(f"review handoff receipt JSON after {context} changed its evidence summary")
    if receipt.get("remainingApprovals") != handoff_value["nextAction"]["approvalRequiredBefore"]:
        raise SystemExit(f"review handoff receipt JSON after {context} changed remaining approvals")
    if receipt.get("nextAction") != {
        "id": "target-repo-intake-required",
        "status": "pending",
        "summary": "Inspect the declared target repository before any implementation begins.",
        "implementationAuthorized": False,
    }:
        raise SystemExit(f"review handoff receipt JSON after {context} changed its next action")
    if receipt.get("boundary") != {
        "mode": "read-only",
        "localWrites": False,
        "targetRepoMutation": False,
        "externalWrites": False,
        "transportVerified": False,
        "consumerIdentityVerified": False,
        "acceptanceRecorded": False,
        "implementationStarted": False,
    }:
        raise SystemExit(f"review handoff receipt JSON after {context} exceeded its proof boundary")
